src/ecmult_gen_impl.h

```diff
@@ -29,38 +29,154 @@ static void secp256k1_ecmult_gen_context_clear(secp256k1_ecmult_gen_context *ctx
+     * left with computing R = (gn-b)*G + ctx->ge_offset.
+     *
+     * The multiplication (gn-b)*G will be performed using a signed-digit multi-comb (see Section
+     * 3.3 of "Fast and compact elliptic-curve cryptography" by Mike Hamburg
+     * (https://eprint.iacr.org/2012/309).
+     *
+     * Let comb(s,P) = sum((2*s_i-1)*2^i*P for i=0..COMB_BITS-1), where s_i is the i'th bit of the
+     * binary representation of scalar s. So the s_i values determine whether -2^i*P (s_i=0) or
+     * +2^i*P (s_1) are added together. By manipulating:
+     *
+     *     comb(s,P) = sum((2*s_i-1)*2^i*P for i=0..COMB_BITS-1)
+     * <=> comb(s,P) = sum((2*s_i-1)*2^i for i=0..COMB_BITS-1) * P
+     * <=> comb(s,P) = (2*sum(s_i*2^i for i=0..COMB_BITS-1) - sum(2^i for i=0..COMB_BITS-1)) * P
+     * <=> comb(s,P) = (2*s - (2^COMB_BITS - 1)) * P
+     *
+     * Thus (gn-b)*G can be written as comb(s,G) if gn-b = 2*s - (2^COMB_BITS - 1), or
+     * s = (gn - b + (2^COMB_BITS - 1))/2 mod order.
+     *
+     * We use an alternative that avoids the modular division by two: we write (gn-b)*G =
+     * comb(d,G/2). For that to hold it must be the case that
+     * (gn-b)*G = (2*d - (2^COMB_BITS - 1)) * (G/2), or
+     * d = (gn + (2^COMB_BITS - 1)/2 - b) mod order.
+     *
+     * -b is precomputed as ctx->scalar_offset, so our final equations become:
+     *
+     *   d = gn + ctx->scalar_offset + (2^COMB_BITS - 1)/2 (mod order)
+     *   R = comb(d, G/2) + ctx->ge_offset.
+     *
+     * The comb function is computed by summing + or - 2^(i-1)*G, for i=0..COMB_BITS-1, depending
+     * on the value of the bits d_i of the binary representation of scalar d.
+     */
+
+    /* Compute the scalar (gn + ctx->scalar_offset + (2^COMB_BITS - 1)/2). */
+    secp256k1_ecmult_gen_scalar_diff(&recoded);
+    secp256k1_scalar_add(&recoded, &recoded, &ctx->scalar_offset);
+    secp256k1_scalar_add(&recoded, &recoded, gn);
+
+    /* In secp256k1_ecmult_gen_prec_table we have precomputed sums of the
+     * (2*d_i-1) * 2^(i-1) * G points, for various combinations of i positions.
+     * We rewrite our equation in terms of these table entries.
+     *
+     * Let mask(b) = sum(2^((b*COMB_TEETH + t)*COMB_SPACING) for t=0..COMB_TEETH-1),
+     * with b ranging from 0 to COMB_BLOCKS-1. So for example with COMB_BLOCKS=11,
+     * COMB_TEETH=6, COMB_SPACING=4, we would have:
+     *   mask(0)  = 2^0   + 2^4   + 2^8   + 2^12  + 2^16  + 2^20,
+     *   mask(1)  = 2^24  + 2^28  + 2^32  + 2^36  + 2^40  + 2^44,
+     *   mask(2)  = 2^48  + 2^52  + 2^56  + 2^60  + 2^64  + 2^68,
+     *   ...
+     *   mask(10) = 2^240 + 2^244 + 2^248 + 2^252 + 2^256 + 2^260
+     *
+     * Imagine we have a table(b,m) function which can look up, given b and
+     * m=(recoded & mask(b)), the sum of (2*d_i-1)*2^(i-1)*G for all bit positions
+     * i set in mask(b). In our example, table(0, 1 + 2^8 + 2^20) would be equal to
+     * (2^-1 - 2^3 + 2^7 - 2^11 - 2^15 + 2^19)*G.
```

Comment:
Hm, this took me a while to follow. Let me explain in my own words. So you assume that the relevant bits of `d` are `1...0...0...1...0...1`, or `1...0...1...0...0...1` in reverse, which is `1 + 2^8 + 2^20`, which determines the positive summands in the G factor (decremented because the formula says so). Is this right? Want to elaborate in the comment?

Reply:
Yes!